File: platform/how-to-get-started/2.fetch/2.gemini-openai-tool-calls.mjs
/**
 * Gemini OpenAI-Compatible Tool Calls Example
 *
 * Uses fetch_undici to demonstrate function calling (tool calls) with
 * Gemini's OpenAI-compatible endpoint.
 *
 * Run: node 2.gemini-openai-tool-calls.mjs
 */
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const ROOT_DIR = join(__dirname, '../..');

const { request } = await import(join(ROOT_DIR, 'packages_mjs/fetch_undici/dist/index.js'));

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
  console.error('Error: GEMINI_API_KEY environment variable is required');
  process.exit(1);
}

// Gemini OpenAI-compatible endpoint
const CHAT_ENDPOINT = 'https://generativelanguage.googleapis.com/v1beta/openai/chat/completions';

async function chatCompletion(messages, options = {}) {
  const payload = {
    model: options.model || 'gemini-2.0-flash',
    messages,
    temperature: options.temperature ?? 0.7,
    max_tokens: options.max_tokens ?? 1000,
    ...options,
  };

  const { statusCode, body } = await request(CHAT_ENDPOINT, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${GEMINI_API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  const text = await body.text();

  if (statusCode >= 400) {
    throw new Error(`${statusCode}: ${text}`);
  }

  return JSON.parse(text);
}

// Define tools in OpenAI format
const weatherTool = {
  type: 'function',
  function: {
    name: 'get_weather',
    description: 'Get the current weather for a location',
    parameters: {
      type: 'object',
      properties: {
        location: {
          type: 'string',
          description: 'The city and state, e.g., San Francisco, CA',
        },
        unit: {
          type: 'string',
          enum: ['celsius', 'fahrenheit'],
          description: 'Temperature unit',
        },
      },
      required: ['location'],
    },
  },
};

const calculatorTool = {
  type: 'function',
  function: {
    name: 'calculate',
    description: 'Perform a mathematical calculation',
    parameters: {
      type: 'object',
      properties: {
        expression: {
          type: 'string',
          description: 'The mathematical expression to evaluate',
        },
      },
      required: ['expression'],
    },
  },
};

// Simulated tool implementations
function executeWeather(args) {
  const { location, unit = 'celsius' } = args;
  // Simulated weather data
  return {
    location,
    temperature: unit === 'celsius' ? 22 : 72,
    unit,
    conditions: 'sunny',
    humidity: 45,
  };
}

function tokenizeMath(expr) {
  const tokens = [];
  let i = 0;
  while (i < expr.length) {
    if (expr[i] === ' ') { i++; continue; }
    if ('+-*/()'.includes(expr[i])) {
      tokens.push(expr[i++]);
    } else if (/[0-9.]/.test(expr[i])) {
      let num = '';
      while (i < expr.length && /[0-9.]/.test(expr[i])) num += expr[i++];
      tokens.push(num);
    } else {
      throw new Error(`Unexpected character: ${expr[i]}`);
    }
  }
  return tokens;
}

function safeEvalMath(expression) {
  const tokens = tokenizeMath(expression);
  let pos = 0;
  const peek = () => tokens[pos];
  const consume = () => tokens[pos++];

  function parseExpr() {
    let left = parseTerm();
    while (peek() === '+' || peek() === '-') {
      const op = consume();
      const right = parseTerm();
      left = op === '+' ? left + right : left - right;
    }
    return left;
  }

  function parseTerm() {
    let left = parseFactor();
    while (peek() === '*' || peek() === '/') {
      const op = consume();
      const right = parseFactor();
      left = op === '*' ? left * right : left / right;
    }
    return left;
  }

  function parseFactor() {
    if (peek() === '(') {
      consume();
      const val = parseExpr();
      if (peek() !== ')') throw new Error('Missing closing parenthesis');
      consume();
      return val;
    }
    if (peek() === '-') { consume(); return -parseFactor(); }
    if (peek() === '+') { consume(); return parseFactor(); }
    const token = consume();
    if (token === undefined) throw new Error('Unexpected end of expression');
    const num = parseFloat(token);
    if (isNaN(num)) throw new Error(`Invalid number: ${token}`);
    return num;
  }

  const result = parseExpr();
  if (pos !== tokens.length) throw new Error('Unexpected token after expression');
  return result;
}

function executeCalculate(args) {
  const { expression } = args;
  try {
    const result = safeEvalMath(expression);
    return { expression, result };
  } catch (e) {
    return { expression, error: 'Invalid expression' };
  }
}

async function main() {
  console.log('='.repeat(60));
  console.log('Gemini OpenAI-Compatible Tool Calls Examples');
  console.log('='.repeat(60));
  console.log();

  // Example 1: Basic tool call
  console.log('--- Example 1: Basic Tool Call ---');
  try {
    const response = await chatCompletion([
      { role: 'system', content: 'You are a helpful assistant. Use the provided tools when appropriate.' },
      { role: 'user', content: 'What is the weather in Tokyo?' }
    ], {
      tools: [weatherTool],
      tool_choice: 'auto',
    });

    console.log('Response:');
    const choice = response.choices[0];
    console.log('Finish reason:', choice.finish_reason);

    if (choice.message.tool_calls) {
      console.log('Tool calls:');
      for (const toolCall of choice.message.tool_calls) {
        console.log(`  - ${toolCall.function.name}(${toolCall.function.arguments})`);

        // Execute the tool
        const args = JSON.parse(toolCall.function.arguments);
        const result = executeWeather(args);
        console.log('  Result:', JSON.stringify(result));
      }
    } else {
      console.log('Content:', choice.message.content);
    }
  } catch (error) {
    console.error('Error:', error.message);
  }
  console.log();

  // Example 2: Multiple tools available
  console.log('--- Example 2: Multiple Tools Available ---');
  try {
    const response = await chatCompletion([
      { role: 'user', content: 'What is 15 * 7 + 23?' }
    ], {
      tools: [weatherTool, calculatorTool],
      tool_choice: 'auto',
    });

    const choice = response.choices[0];
    if (choice.message.tool_calls) {
      console.log('Tool calls:');
      for (const toolCall of choice.message.tool_calls) {
        console.log(`  - ${toolCall.function.name}(${toolCall.function.arguments})`);

        const args = JSON.parse(toolCall.function.arguments);
        if (toolCall.function.name === 'calculate') {
          const result = executeCalculate(args);
          console.log('  Result:', JSON.stringify(result));
        }
      }
    } else {
      console.log('Content:', choice.message.content);
    }
  } catch (error) {
    console.error('Error:', error.message);
  }
  console.log();

  // Example 3: Complete tool call conversation
  console.log('--- Example 3: Complete Tool Call Conversation ---');
  try {
    // Initial request
    const response1 = await chatCompletion([
      { role: 'user', content: 'What is the weather in Paris?' }
    ], {
      tools: [weatherTool],
    });

    const choice1 = response1.choices[0];
    const toolCalls = choice1.message.tool_calls;

    if (toolCalls) {
      console.log('Step 1 - Model requested tool call:');
      console.log(`  ${toolCalls[0].function.name}(${toolCalls[0].function.arguments})`);

      // Execute tool and get result
      const args = JSON.parse(toolCalls[0].function.arguments);
      const toolResult = executeWeather(args);
      console.log('Step 2 - Tool result:', JSON.stringify(toolResult));

      // Continue conversation with tool result
      const response2 = await chatCompletion([
        { role: 'user', content: 'What is the weather in Paris?' },
        {
          role: 'assistant',
          content: null,
          tool_calls: toolCalls,
        },
        {
          role: 'tool',
          tool_call_id: toolCalls[0].id,
          content: JSON.stringify(toolResult),
        },
      ], {
        tools: [weatherTool],
      });

      console.log('Step 3 - Final response:');
      console.log(response2.choices[0].message.content);
    }
  } catch (error) {
    console.error('Error:', error.message);
  }
  console.log();

  // Example 4: Forced tool choice
  console.log('--- Example 4: Forced Tool Choice ---');
  try {
    const response = await chatCompletion([
      { role: 'user', content: 'Tell me about the Eiffel Tower.' }
    ], {
      tools: [weatherTool],
      tool_choice: { type: 'function', function: { name: 'get_weather' } },
    });

    const choice = response.choices[0];
    if (choice.message.tool_calls) {
      console.log('Forced tool call:');
      for (const toolCall of choice.message.tool_calls) {
        console.log(`  - ${toolCall.function.name}(${toolCall.function.arguments})`);
      }
    }
  } catch (error) {
    console.error('Error:', error.message);
  }
  console.log();

  // Example 5: Tool definition format
  console.log('--- Example 5: OpenAI Tool Definition Format ---');
  console.log(`
┌─────────────────────────────────────────────────────────────────┐
│                   OPENAI TOOL DEFINITION FORMAT                  │
├─────────────────────────────────────────────────────────────────┤
│ {                                                               │
│   "type": "function",                                           │
│   "function": {                                                 │
│     "name": "function_name",                                    │
│     "description": "What the function does",                    │
│     "parameters": {                                             │
│       "type": "object",                                         │
│       "properties": {                                           │
│         "param1": { "type": "string", "description": "..." },   │
│         "param2": { "type": "number", "description": "..." }    │
│       },                                                        │
│       "required": ["param1"]                                    │
│     }                                                           │
│   }                                                             │
│ }                                                               │
├─────────────────────────────────────────────────────────────────┤
│ Tool Choice Options:                                            │
│ • "auto" - Model decides whether to use tools                   │
│ • "none" - Model should not use any tools                       │
│ • "required" - Model must use a tool                            │
│ • {"type":"function","function":{"name":"X"}} - Force specific  │
└─────────────────────────────────────────────────────────────────┘
`);

  console.log('='.repeat(60));
  console.log('Tool Calls Examples Complete');
  console.log('='.repeat(60));
}

main().catch(console.error);

File: platform/how-to-get-started/2.fetch/2.gemini-openai-tool-calls.py
#!/usr/bin/env python3
"""
Gemini OpenAI-Compatible Tool Calls Example

Uses fetch_httpx to demonstrate function calling (tool calls) with
Gemini's OpenAI-compatible endpoint.

Run: python 2.gemini-openai-tool-calls.py
"""
import ast
import asyncio
import json
import operator
import os
import sys
from pathlib import Path

# Add packages to path
ROOT_DIR = Path(__file__).parent.parent.parent
sys.path.insert(0, str(ROOT_DIR / "packages_py"))

from fetch_httpx import AsyncClient

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
if not GEMINI_API_KEY:
    print("Error: GEMINI_API_KEY environment variable is required")
    sys.exit(1)

# Gemini OpenAI-compatible endpoint
CHAT_ENDPOINT = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions"


async def chat_completion(messages, **options):
    payload = {
        "model": options.get("model", "gemini-2.0-flash"),
        "messages": messages,
        "temperature": options.get("temperature", 0.7),
        "max_tokens": options.get("max_tokens", 1000),
        **{k: v for k, v in options.items() if k not in ["model", "temperature", "max_tokens"]},
    }

    async with AsyncClient(timeout=60.0) as client:
        response = await client.post(
            CHAT_ENDPOINT,
            headers={
                "Authorization": f"Bearer {GEMINI_API_KEY}",
                "Content-Type": "application/json",
            },
            json=payload,
        )

        if response.status_code >= 400:
            raise Exception(f"{response.status_code}: {response.text}")

        return response.json()


# Define tools in OpenAI format
weather_tool = {
    "type": "function",
    "function": {
        "name": "get_weather",
        "description": "Get the current weather for a location",
        "parameters": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "The city and state, e.g., San Francisco, CA",
                },
                "unit": {
                    "type": "string",
                    "enum": ["celsius", "fahrenheit"],
                    "description": "Temperature unit",
                },
            },
            "required": ["location"],
        },
    },
}

calculator_tool = {
    "type": "function",
    "function": {
        "name": "calculate",
        "description": "Perform a mathematical calculation",
        "parameters": {
            "type": "object",
            "properties": {
                "expression": {
                    "type": "string",
                    "description": "The mathematical expression to evaluate",
                },
            },
            "required": ["expression"],
        },
    },
}


# Simulated tool implementations
def execute_weather(args):
    location = args.get("location")
    unit = args.get("unit", "celsius")
    # Simulated weather data
    return {
        "location": location,
        "temperature": 22 if unit == "celsius" else 72,
        "unit": unit,
        "conditions": "sunny",
        "humidity": 45,
    }


_SAFE_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}


def _eval_math_node(node):
    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
        return node.value
    if isinstance(node, ast.BinOp):
        op = _SAFE_OPS.get(type(node.op))
        if op is None:
            raise ValueError(f"Unsupported operator: {type(node.op).__name__}")
        return op(_eval_math_node(node.left), _eval_math_node(node.right))
    if isinstance(node, ast.UnaryOp):
        op = _SAFE_OPS.get(type(node.op))
        if op is None:
            raise ValueError(f"Unsupported operator: {type(node.op).__name__}")
        return op(_eval_math_node(node.operand))
    raise ValueError(f"Unsupported expression: {type(node).__name__}")


def safe_eval_math(expression):
    tree = ast.parse(expression.strip(), mode="eval")
    return _eval_math_node(tree.body)


def execute_calculate(args):
    expression = args.get("expression")
    try:
        result = safe_eval_math(expression)
        return {"expression": expression, "result": result}
    except Exception:
        return {"expression": expression, "error": "Invalid expression"}


async def main():
    print("=" * 60)
    print("Gemini OpenAI-Compatible Tool Calls Examples")
    print("=" * 60)
    print()

    # Example 1: Basic tool call
    print("--- Example 1: Basic Tool Call ---")
    try:
        response = await chat_completion(
            [
                {"role": "system", "content": "You are a helpful assistant. Use the provided tools when appropriate."},
                {"role": "user", "content": "What is the weather in Tokyo?"}
            ],
            tools=[weather_tool],
            tool_choice="auto",
        )

        print("Response:")
        choice = response["choices"][0]
        print(f"Finish reason: {choice['finish_reason']}")

        tool_calls = choice["message"].get("tool_calls")
        if tool_calls:
            print("Tool calls:")
            for tool_call in tool_calls:
                print(f"  - {tool_call['function']['name']}({tool_call['function']['arguments']})")

                # Execute the tool
                args = json.loads(tool_call["function"]["arguments"])
                result = execute_weather(args)
                print(f"  Result: {json.dumps(result)}")
        else:
            print(f"Content: {choice['message'].get('content')}")
    except Exception as e:
        print(f"Error: {e}")
    print()

    # Example 2: Multiple tools available
    print("--- Example 2: Multiple Tools Available ---")
    try:
        response = await chat_completion(
            [{"role": "user", "content": "What is 15 * 7 + 23?"}],
            tools=[weather_tool, calculator_tool],
            tool_choice="auto",
        )

        choice = response["choices"][0]
        tool_calls = choice["message"].get("tool_calls")
        if tool_calls:
            print("Tool calls:")
            for tool_call in tool_calls:
                print(f"  - {tool_call['function']['name']}({tool_call['function']['arguments']})")

                args = json.loads(tool_call["function"]["arguments"])
                if tool_call["function"]["name"] == "calculate":
                    result = execute_calculate(args)
                    print(f"  Result: {json.dumps(result)}")
        else:
            print(f"Content: {choice['message'].get('content')}")
    except Exception as e:
        print(f"Error: {e}")
    print()

    # Example 3: Complete tool call conversation
    print("--- Example 3: Complete Tool Call Conversation ---")
    try:
        # Initial request
        response1 = await chat_completion(
            [{"role": "user", "content": "What is the weather in Paris?"}],
            tools=[weather_tool],
        )

        choice1 = response1["choices"][0]
        tool_calls = choice1["message"].get("tool_calls")

        if tool_calls:
            print("Step 1 - Model requested tool call:")
            print(f"  {tool_calls[0]['function']['name']}({tool_calls[0]['function']['arguments']})")

            # Execute tool and get result
            args = json.loads(tool_calls[0]["function"]["arguments"])
            tool_result = execute_weather(args)
            print(f"Step 2 - Tool result: {json.dumps(tool_result)}")

            # Continue conversation with tool result
            response2 = await chat_completion(
                [
                    {"role": "user", "content": "What is the weather in Paris?"},
                    {
                        "role": "assistant",
                        "content": None,
                        "tool_calls": tool_calls,
                    },
                    {
                        "role": "tool",
                        "tool_call_id": tool_calls[0]["id"],
                        "content": json.dumps(tool_result),
                    },
                ],
                tools=[weather_tool],
            )

            print("Step 3 - Final response:")
            print(response2["choices"][0]["message"]["content"])
    except Exception as e:
        print(f"Error: {e}")
    print()

    # Example 4: Forced tool choice
    print("--- Example 4: Forced Tool Choice ---")
    try:
        response = await chat_completion(
            [{"role": "user", "content": "Tell me about the Eiffel Tower."}],
            tools=[weather_tool],
            tool_choice={"type": "function", "function": {"name": "get_weather"}},
        )

        choice = response["choices"][0]
        tool_calls = choice["message"].get("tool_calls")
        if tool_calls:
            print("Forced tool call:")
            for tool_call in tool_calls:
                print(f"  - {tool_call['function']['name']}({tool_call['function']['arguments']})")
    except Exception as e:
        print(f"Error: {e}")
    print()

    # Example 5: Tool definition format
    print("--- Example 5: OpenAI Tool Definition Format ---")
    print("""
┌─────────────────────────────────────────────────────────────────┐
│                   OPENAI TOOL DEFINITION FORMAT                  │
├─────────────────────────────────────────────────────────────────┤
│ {                                                               │
│   "type": "function",                                           │
│   "function": {                                                 │
│     "name": "function_name",                                    │
│     "description": "What the function does",                    │
│     "parameters": {                                             │
│       "type": "object",                                         │
│       "properties": {                                           │
│         "param1": { "type": "string", "description": "..." },   │
│         "param2": { "type": "number", "description": "..." }    │
│       },                                                        │
│       "required": ["param1"]                                    │
│     }                                                           │
│   }                                                             │
│ }                                                               │
├─────────────────────────────────────────────────────────────────┤
│ Tool Choice Options:                                            │
│ • "auto" - Model decides whether to use tools                   │
│ • "none" - Model should not use any tools                       │
│ • "required" - Model must use a tool                            │
│ • {"type":"function","function":{"name":"X"}} - Force specific  │
└─────────────────────────────────────────────────────────────────┘
""")

    print("=" * 60)
    print("Tool Calls Examples Complete")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())
